main.cpp

```diff
@@ -230,63 +158,62 @@ struct IntType;
 struct DoubleType;
 
 /*-------------------------------------------------------------------------------*/
-template<typename numType>
+template<typename NumericType>
 struct Numeric
 {
-    using Type = numType;
+    using Type = Temporary<NumericType>;
 private:
     std::unique_ptr<Type> value;
-    Numeric& powInternal(Type x)
-    {
-        *value = static_cast<Type>(std::pow( *value, x ));
-        return *this;
-    }
 public:
     Numeric( Type x ) : value( std::make_unique<Type>(x) ) { }
 
-    Numeric& apply( std::function<Numeric&( std::unique_ptr<Type>& )> f )
+    template<typename Callable>
+    Numeric& apply( Callable&& f )
     {
-        if( f != nullptr )
-        {
-            f( value );
-        }
+        f( value );
+
         return *this;
-    } 
-    
-    Numeric& apply( void( *f )( std::unique_ptr<Type>& ) )
+    }
+
+    template<typename OtherType>
+    Numeric& pow(const OtherType& x)
     {
-        if( f != nullptr )
-        {
-            f( value );
-        }
+        *value = static_cast<NumericType>(std::pow( *value, static_cast<NumericType>(x) ));
         return *this;
     }
 
-    Numeric& pow(const Numeric& x)
+    template<typename OtherType>
+    Numeric& operator=( const OtherType& x)
     {
-        return powInternal( x );
+        *value = static_cast<NumericType>(x);
+        return *this;
     }
 
-    Numeric& operator+=( Type x )
+    template<typename OtherType>
+    Numeric& operator+=( const OtherType& x )
     {
-        *value += x;
+        *value += static_cast<NumericType>(x);
         return *this;
     }
-    Numeric& operator-=( Type x )
+
+    template<typename OtherType>
+    Numeric& operator-=( const OtherType& x )
     {
-        *value -= x;
+        *value -= static_cast<NumericType>(x);
         return *this;
     }
-    Numeric& operator*=( Type x )
+
+    template<typename OtherType>
+    Numeric& operator*=( const OtherType& x )
     {
-        *value *= x;
+        *value *= static_cast<NumericType>(x);
         return *this;
     }
     
     template<typename DivisorType>
-    Numeric& operator/=( DivisorType x )
+    Numeric& operator/=( const DivisorType& x )
     {
-        if( std::is_same<int, Type>::value )
+        if( std::is_same<int, NumericType>::value )
         {
             if( std::is_same<int, DivisorType>::value )
             {
```

Comment:
for the line below this one, if you make it into this block, you already know that `DivisorType` is `int`
If you're getting a warning without this cast, it's because you didn't make these expressions `constexpr`
- [ ] revise